### PR TITLE
Move creation of FSTFilter objects to static create method on FSTFilter.

### PR DESCRIPTION
**Pull request by rsgowman**
Rather than previously inlining it in the calling code. This is to unify
filter creation across platforms.

(This change involves altering FSTFilter from a protocol to an abstract
class.)

### CODE REVIEW

File: Firestore/Source/Core/FSTQuery.mm [hunk 1]
```diff
@@ -66,6 +68,43 @@ static constexpr NSComparisonResult ReverseOrder(NSComparisonResult result) {
...
+                           @"Invalid Query. You can only perform equality comparisons on NaN.");
+    }
+    return [[FSTNanFilter alloc] initWithField:field];
+  } else {
+    return [FSTRelationFilter filterWithField:field filterOperator:op value:value];
+  }
+}
+
+- (const FieldPath &)field {
+  @throw FSTAbstractMethodException();  // NOLINT
```
Comment (rsgowman) [inline Firestore/Source/Core/FSTQuery.mm]: The docs for FSTAbstractMethodException suggest that adding `// NOLINT` is necessary... but a quick run of the linter without it seemed to pass.

After this PR is submitted, I'll check that a bit more thoroughly, and if it turns out to not be necessary, I'll remove it everywhere.

Comment (wilhuff) [reply Firestore/Source/Core/FSTQuery.mm]: The Google-internal objective-c linter is the one that complains. We've never managed to integrate it into our Objective-C build on github.

File: Firestore/Source/Core/FSTQuery.h [hunk 2]
```diff
@@ -38,15 +38,27 @@ typedef NS_ENUM(NSInteger, FSTRelationFilterOperator) {
...
-/** Returns the field the Filter operates over. */
+/**
+ * Creates a filter for the provided path, operator, and value.
+ *
+ * Note that if the relational operator is FSTRelationFilterOperatorEqual and
+ * the value is [FSTNullValue nullValue] or [FSTDoubleValue nanValue], this
+ * will return the appropriate FSTNullFilter or FSTNanFilter class instead of a
+ * FSTRelationFilter.
+ */
++ (instancetype)createWithField:(const firebase::firestore::model::FieldPath &)field
```
Comment (wilhuff) [inline Firestore/Source/Core/FSTQuery.h]: Objective-C functions that create a thing name it as a first part of the label. There's usually no explicit "create" or "make" (e.g. `NSString stringWithFormat:`). This should therefore start with "filter" (as does FSTRelationFilter's factory).

The labels here should match the ones in `FSTRelationFilter`. This is `...WithField:op:value`, that one is `...WithField:filterOperator:value`. Objective-C style favors plain language over abbreviation, so probably filterOperator is better. (This used to be just "operator" but that works in Objective-C but not Objective-C++ where "operator" is a reserved word. This is one of the few areas where Objective-C++ is not a strict superset of Objective-C.)

File: Firestore/Source/Core/FSTQuery.h [hunk 3]
```diff
@@ -55,7 +67,7 @@ typedef NS_ENUM(NSInteger, FSTRelationFilterOperator) {
  * FSTRelationFilter is a document filter constraint on a query with a single relation operator.
  * It is similar to NSComparisonPredicate, except customized for Firestore semantics.
  */
-@interface FSTRelationFilter : NSObject <FSTFilter>
+@interface FSTRelationFilter : FSTFilter
```
Comment (wilhuff) [inline Firestore/Source/Core/FSTQuery.h]: We can likely convert the factory function on `FSTRelationFilter` to just be an initializer (i.e. rename to `initWith...` and make non-static). Callers then call `[[FSTRelationFilter alloc] initWithField:...]

File: Firestore/Source/Core/FSTQuery.h (hunk 2 above)
Comment (rsgowman) [reply Firestore/Source/Core/FSTQuery.h]: Done.

Fun fact: Actually using this method (eg api/firquery.mm) uses the word "filter" 6 times in a single line.
```
FSTFilter *filter = [FSTFilter filterWithField:fieldPath filterOperator:filterOperator value:field];
```

File: Firestore/Source/Core/FSTQuery.h (hunk 3 above)
Comment (rsgowman) [reply Firestore/Source/Core/FSTQuery.h]: Done.

File: Firestore/Source/Core/FSTQuery.mm (hunk 1 above)
Comment (rsgowman) [reply Firestore/Source/Core/FSTQuery.mm]: Ah, interesting. I'll leave it as is then.

File: Firestore/Source/Core/FSTQuery.h (hunk 2 above)
Comment (wilhuff) [reply Firestore/Source/Core/FSTQuery.h]: It wouldn't be Objective-C if you weren't repeating yourself at least three times.

**Pull request merged**